### PR TITLE
feat: reducer max for `axis==None`

### PR DESCRIPTION
**Pull request by ianna**

### CODE REVIEW

**Comment by ikrommyd** [comment]
Why is `maximum` needed? Isn't `initial` only used when the input is empty? That's already covered by the `if data.size == 0` branch no?

**Comment by ianna** [comment]
> Why is `maximum` needed? Isn't `initial` only used when the input is empty? That's already covered by the `if data.size == 0` branch no?

No. `initial` is the value with which we start the reduction.

`maximum` is needed to avoid device to host copy on GPU.

**Comment by ikrommyd** [comment]
FYI
```
In [1]: import numpy as np

In [2]: import awkward as ak

In [3]: ak.jax.register_and_check()

In [4]: x = ak.Array(np.array([]), backend="jax")
/home/iason/Dropbox/work/pyhep_dev/awkward/src/awkward/_singleton.py:19: DeprecationWarning: The JAX backend is deprecated and will be removed in a future release of Awkward Array. Please plan to migrate your code accordingly.
  self.__init__()  # pylint: disable=unnecessary-dunder-call

In [5]: ak.max(x)
[1]    12443 segmentation fault (core dumped)  ipython
```
Nothing to do with this pr branch, it segfaults on main too.

Edit: oh this is when jax automatically puts arrays on GPU
```
In [5]: x.layout.data
Out[5]: Array([], shape=(0,), dtype=float64)

In [6]: x.layout.data.device
Out[6]: CudaDevice(id=0)
```
Maybe we should automatically force cpu for jax when we do register and check? Cause as far as I know we pass jax to the awkward-cpp kernels too and that can't work if the arrays are on gpu.

File: src/awkward/_connect/jax/reducers.py [hunk 1]
```diff
@@ -656,5 +703,42 @@ def apply(
...
+        parents: ak.index.Index,
+        offsets: ak.index.Index | ak.index.EmptyIndex,
+        starts: ak.index.Index,
+        shifts: ak.index.Index | None,
+        outlength: ShapeItem,
+    ) -> ak.contents.NumpyArray:
+        data = maybe_materialize(array.data)[0]
+
+        # For axis=None, just use JAX's native max
+        result_scalar = jax.numpy.max(data)
```
Comment (ikrommyd) [inline src/awkward/_connect/jax/reducers.py]: You'd get an error here if `data` is empty. Can that happen?

Comment (ianna) [reply src/awkward/_connect/jax/reducers.py]: Good catch!

**Comment by ianna** [comment]
> FYI
> 
> ```
> In [1]: import numpy as np
> 
> In [2]: import awkward as ak
> 
> In [3]: ak.jax.register_and_check()
> 
> In [4]: x = ak.Array(np.array([]), backend="jax")
> /home/iason/Dropbox/work/pyhep_dev/awkward/src/awkward/_singleton.py:19: DeprecationWarning: The JAX backend is deprecated and will be removed in a future release of Awkward Array. Please plan to migrate your code accordingly.
>   self.__init__()  # pylint: disable=unnecessary-dunder-call
> 
> In [5]: ak.max(x)
> [1]    12443 segmentation fault (core dumped)  ipython
> ```
> 
> Nothing to do with this pr branch, it segfaults on main too.
> 
> Edit: oh this is when jax automatically puts arrays on GPU
> 
> ```
> In [5]: x.layout.data
> Out[5]: Array([], shape=(0,), dtype=float64)
> 
> In [6]: x.layout.data.device
> Out[6]: CudaDevice(id=0)
> ```
> 
> Maybe we should automatically force cpu for jax when we do register and check? Cause as far as I know we pass jax to the awkward-cpp kernels too and that can't work if the arrays are on gpu.

Thanks. I should check it on GPU next. I think, it should be ok as is for now. When there is no specific JAX kernel we fall back on CPU, but if any of this is a bottleneck we can provide an efficient way to switch to either CUDA or `cuda.compute`.

**Comment by ikrommyd** [comment]
Yeah I think it's trying to run a CPU kernel when the buffer is actually on GPU and segfaulting.

**Comment by ianna** [comment]
> Yeah I think it's trying to run a CPU kernel when the buffer is actually on GPU and segfaulting.

can you tell which kernel?
